File: tools/capture_phc04_legislative_consequence.mjs
#!/usr/bin/env node
/**
 * PHC-04 evidence helper: render the meeting detail page (site/meeting_document.mjs's
 * renderMeetingDocument()) for five synthetic City Record records, one per state named
 * in the card (a single exact matter join, several strict matches, an adversarial
 * title-only near-match, a later published vote, and a hearing that is itself a
 * matter's most recent appearance). Writes each rendered document under
 * site/.phc04-capture-tmp/ so tools/capture_phc04_legislative_consequence.py can serve
 * it locally (real /brand.css, /civic-documents.css, /report_issue.mjs) and run axe-core
 * against it. Prints the case manifest (id, path, assertion) as JSON to stdout.
 *
 *   node tools/capture_phc04_legislative_consequence.mjs
 */
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";

import { renderMeetingDocument } from "../site/meeting_document.mjs";

const ROOT = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const OUT_DIR = path.join(ROOT, "site/.phc04-capture-tmp");

const goldFixtures = JSON.parse(
  readFileSync(path.join(ROOT, "test/fixtures/consequence_projection/gold_fixtures.v0.json"), "utf8"),
);
const goldCase = (id) => goldFixtures.cases.find((c) => c.id === id).record;

function fullCouncilRecord(overrides = {}) {
  const record = structuredClone(goldCase("full-council-hearing-land-use-matched-join"));
  record.agency = "City Council";
  return { ...record, ...overrides };
}

function present(matters) {
  return {
    snapshot_state: "present",
    join: { matched: true, method: "exact_date_body_tokens" },
    matters,
  };
}

// Matter 78605 (LU 0056-2026) has two observed appearances in
// site/data/legislative_matter_lookup.json: a 2026-04-22 hearing that laid it
// over, then a 2026-05-19 hearing that recorded a roll-call vote approving it.
function matter78605Record(requestId, { outcome, eventName, eventDate }) {
  return {
    source_system: "city_record",
    meeting_id: `meeting:city_record:${requestId}`,
    request_id: requestId,
    event_date: `${eventDate}T11:00:00-04:00`,
    decides: "Zoning, 147-14 Northern Boulevard Rezoning, Queens (C 220415 ZMQ).",
    council_hearing_kind: "land_use",
    meeting_outcome: {
      request_id: requestId,
      snapshot_state: "present",
      event: { event_id: "any", name: eventName, date: eventDate, url: "https://nyc.legistar.com/MeetingDetail.aspx?LEGID=any", documents: [] },
      matters: [{
        matter_id: "78605",
        matter_file: "LU 0056-2026",
        matter_url: "https://nyc.legistar.com/Gateway.aspx?M=L&ID=78605",
        title: "Zoning, 147-14 Northern Boulevard Rezoning, Queens (C 220415 ZMQ).",
        outcome,
      }],
      join: { matched: true, method: "exact_date_body_tokens" },
    },
  };
}

const CASES = [
  {
    id: "single_exact_match_process_and_testimony",
    assertion:
      "A1/A4/A5/G1: a single exact matter join renders a consequence block above the existing " +
      "continuation, naming the committee's own recorded process position and where testimony is " +
      "officially recorded, with public testimony and committee questioning described as separate " +
      "activities.",
    record: fullCouncilRecord({
      meeting_id: "meeting:city_record:phc04-cap-single",
      request_id: "phc04-cap-single",
    }),
  },
  {
    id: "several_strict_matches_no_blended_consequence",
    assertion:
      "A2/G2: a request with several strict matches shows no blended consequence block at all — " +
      "the existing continuation still requires the reader to select a matter.",
    record: {
      source_system: "city_record",
      meeting_id: "meeting:city_record:phc04-cap-multiple",
      request_id: "phc04-cap-multiple",
      event_date: "2026-07-21T11:00:00-04:00",
      meeting_outcome: present([
        { matter_id: "79301", matter_file: "LU 0201-2026", title: "Rezoning A", outcome: "Laid Over by Subcommittee" },
        { matter_id: "79302", matter_file: "LU 0202-2026", title: "Rezoning B", outcome: "Laid Over by Subcommittee" },
      ]),
    },
  },
  {
    id: "adversarial_title_only_near_match_mints_nothing",
    assertion:
      "A3/G3: a title-only, non-strict relation renders no consequence block and leaks no matter " +
      "detail, covered as the adversarial near-match case.",
    record: {
      source_system: "city_record",
      meeting_id: "meeting:city_record:phc04-cap-titleonly",
      request_id: "phc04-cap-titleonly",
      event_date: "2026-07-21T11:00:00-04:00",
      meeting_outcome: {
        snapshot_state: "present",
        join: { matched: true, method: "title_similarity" },
        matters: [{ matter_id: "99999", title: "Same words as the hearing", outcome: "Approved by Subcommittee" }],
      },
    },
  },
  {
    id: "later_published_vote_is_next_official_event",
    assertion:
      "A6: a matter whose lookup history carries a later vote surfaces that vote as the next " +
      "official event, never as any individual comment's effect.",
    record: matter78605Record("20260408025", {
      outcome: "Laid Over by Subcommittee",
      eventName: "Subcommittee on Zoning and Franchises",
      eventDate: "2026-04-22",
    }),
  },
  {
    id: "hearing_is_matters_own_latest_appearance_honest_unknown",
    assertion:
      "A6/negative rule: a hearing that is itself a matter's most recent observed appearance — " +
      "even though its own recorded action was a vote — states honestly that no amendment or vote " +
      "has followed yet, rather than mistaking its own action for a future event.",
    record: matter78605Record("20260428021", {
      outcome: "Approved by Subcommittee",
      eventName: "Subcommittee on Zoning and Franchises",
      eventDate: "2026-05-19",
    }),
  },
];

mkdirSync(OUT_DIR, { recursive: true });
const manifestCases = CASES.map(({ id, assertion, record }) => {
  const html = renderMeetingDocument(record);
  const file = path.join(OUT_DIR, `${id}.html`);
  writeFileSync(file, html, "utf8");
  return { id, assertion, path: `/.phc04-capture-tmp/${id}.html`, meeting_id: record.meeting_id };
});

process.stdout.write(JSON.stringify(manifestCases, null, 2));

File: tools/capture_phc04_legislative_consequence.py
#!/usr/bin/env python3
"""PHC-04 evidence: render the meeting detail page's new "What happens with this matter"
consequence block (site/meeting_document.mjs's renderMeetingDocument(), via
site/legislative_hearing_consequence.mjs) for five synthetic City Record records, one per
state named in the card — a single exact matter join, several strict matches, an
adversarial title-only near-match, a later published vote, and a hearing that is itself a
matter's most recent observed appearance. Runs axe-core against each rendered document at
both review widths.

No image is written or committed — see docs/evidence/public-input-consequence/
legislative-consequence/capture-manifest.json for the committed proof (content hash
+ textual assertion per case), per the workstream's no-committed-image-binaries convention
(spec.md).

  node tools/capture_phc04_legislative_consequence.mjs   # (invoked by this script)
  python3 tools/capture_phc04_legislative_consequence.py
"""

from __future__ import annotations

import hashlib
import json
import shutil
import subprocess
import sys
import tempfile
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "test/functional/assets"))
from a11y_gate import failing_violations  # noqa: E402

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parents[1]
AXE = ROOT / "test/functional/assets/axe.min.js"
VIEWPORTS = (("mobile", 390, 844), ("desktop", 1440, 900))
TMP_DIR = ROOT / "site/.phc04-capture-tmp"
OUT_DIR = ROOT / "docs/evidence/public-input-consequence/legislative-consequence"


def start_local_site_server() -> tuple[subprocess.Popen, str]:
    ready_file = Path(tempfile.mktemp(prefix="crol-phc04-site-"))
    proc = subprocess.Popen(
        [sys.executable, str(ROOT / "tools/local_site_server.py"),
         "--directory", "site", "--port", "0", "--ready-file", str(ready_file)],
        cwd=ROOT,
    )
    deadline = time.time() + 20
    while time.time() < deadline:
        if ready_file.exists() and ready_file.stat().st_size:
            base = ready_file.read_text().strip()
            if base:
                return proc, base
        if proc.poll() is not None:
            raise RuntimeError("local_site_server.py exited before becoming ready")
        time.sleep(0.1)
    raise RuntimeError("timed out waiting for local_site_server.py")


def main() -> None:
    if TMP_DIR.exists():
        shutil.rmtree(TMP_DIR)
    generated = subprocess.run(
        [shutil.which("node") or "node", str(ROOT / "tools/capture_phc04_legislative_consequence.mjs")],
        cwd=ROOT, capture_output=True, text=True, check=True,
    )
    cases = json.loads(generated.stdout)

    server_proc, base = start_local_site_server()
    base = base.rstrip("/")
    captures = []
    try:
        with sync_playwright() as p:
            browser = p.chromium.launch()
            for viewport_name, width, height in VIEWPORTS:
                page = browser.new_page(viewport={"width": width, "height": height})
                for case in cases:
                    page.goto(f"{base}{case['path']}", wait_until="domcontentloaded")
                    page.wait_for_timeout(200)
                    consequence = page.locator('[data-legislative-hearing-consequence="1"]')
                    has_consequence = consequence.count() > 0
                    document_html = page.locator("main#main").inner_html()
                    content_sha256 = hashlib.sha256(document_html.encode("utf-8")).hexdigest()

                    page.add_script_tag(path=str(AXE))
                    result = page.evaluate(
                        "async () => await axe.run(document.querySelector('main#main') || document, "
                        "{resultTypes:['violations']})"
                    )
                    wcag22_rules = set(page.evaluate("() => axe.getRules(['wcag22aa']).map(r => r.ruleId)"))
                    gate = failing_violations(result["violations"], wcag22_rules)

                    captures.append({
                        "fixture": case["id"],
                        "route": f"renderMeetingDocument(record) — {case['meeting_id']}",
                        "viewport": viewport_name,
                        "assertion": case["assertion"],
                        "content_sha256": content_sha256,
                        "consequence_block_present": has_consequence,
                        "axe_critical_or_serious_violations": [v["id"] for v in gate],
                        "file": None,
                    })
                    print(f"{viewport_name:8s} {case['id']:55s} consequence={has_consequence!s:5} "
                          f"sha256={content_sha256[:12]} violations={[v['id'] for v in gate]}")
                page.close()
            browser.close()
    finally:
        server_proc.terminate()
        try:
            server_proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            server_proc.kill()
        shutil.rmtree(TMP_DIR, ignore_errors=True)

    repository_revision = subprocess.run(
        ["git", "rev-parse", "HEAD"], cwd=ROOT, capture_output=True, text=True, check=True,
    ).stdout.strip()

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    manifest = {
        "schema_version": 1,
        "card": "cityscroll-engineering/legislative-consequence-on-exact-joins",
        "capture_mode": "local_node_playwright_axe_static_render",
        "repository_revision": repository_revision,
        "note": (
            "renderMeetingDocument() rendered locally via node (no live Worker/D1 backend) for "
            "five synthetic City Record records, one per state named in the card, then served "
            "from site/ (real /brand.css, /civic-documents.css, /report_issue.mjs) via "
            "tools/local_site_server.py. axe-core (test/functional/assets/axe.min.js) ran "
            "against the rendered <main id=\"main\"> for each case at both review widths, using "
            "the same failing-violation classification as test/functional/11_accessibility.py "
            "(a11y_gate.py). Evidence proves the sourced consequence block's placement, content, "
            "and omission states (A1-A6) render as specified and introduce no new axe violation; "
            "it does not evidence a live-data-populated page, since no Worker/D1 backend was "
            "running."
        ),
        "data_vintage": "not_applicable_synthetic_fixture_records",
        "viewports": [
            {"name": "mobile", "width": 390, "height": 844},
            {"name": "desktop", "width": 1440, "height": 900},
        ],
        "captures": captures,
        "accessibility": {
            "tool": "axe-core (test/functional/assets/axe.min.js)",
            "scope": 'main#main',
            "viewports_checked": ["390x844", "1440x900"],
            "new_violations_introduced": sum(len(c["axe_critical_or_serious_violations"]) for c in captures),
            "pre_existing_violations_observed": [],
        },
    }
    (OUT_DIR / "capture-manifest.json").write_text(
        json.dumps(manifest, indent=2) + "\n", encoding="utf-8"
    )
    print(f"wrote {OUT_DIR / 'capture-manifest.json'}")


if __name__ == "__main__":
    main()
